src/frontend/src/expr/parameter.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+use std::sync::{Arc, RwLock};
+
+use risingwave_common::types::DataType;
+
+use super::Expr;
+
+#[derive(Clone, Debug)]
+pub struct Parameter {
+    pub index: u64,
+    param_types: Arc<RwLock<HashMap<u64, Option<DataType>>>>,
+}
+
+impl PartialEq for Parameter {
+    fn eq(&self, other: &Self) -> bool {
+        self.index == other.index
+    }
+}
+
+impl Eq for Parameter {}
+
+impl Hash for Parameter {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.index.hash(state);
+        self.param_types
+            .read()
+            .unwrap()
+            .get(&self.index)
+            .unwrap()
+            .hash(state);
+    }
+}
+
+impl Expr for Parameter {
+    fn return_type(&self) -> DataType {
+        self.param_types
+            .read()
+            .unwrap()
+            .get(&self.index)
+            .unwrap()
+            .clone()
+            .unwrap_or(DataType::Varchar)
```

Comment:
Is this the behavior of pg that sets default data type to varchar?

Reply:
Yes.